C2LC-290: Worlds menu.

the-t-in-rtf opened the pull request:
See [C2LC-290](https://issues.fluidproject.org/browse/C2LC-290) for details.

the-t-in-rtf commented:
This will not be ready to review until issues inherited from the work on C2LC-286 are resolved (the position of the character is not correct at the moment).

We also need to discuss where the menu should actually live.  I put it where it used to be until we can review and discuss.

chosww commented:
Hey Tony, could you please merge in the latest 286 into this work? Changes to move the heading to the top left corner have some updates to overall layout template, so it makes sense for me to give you feedback related to the layout work with the changes merged in; however, I will still leave some comments about the layout work you did to explain why overall layout in current build site is broken.

chosww commented on src/App.scss:
```diff
@@ -73,84 +104,109 @@ body {
 }
 
-.App__playAndShare-container {
+.App__program-block-editor {
```
I see you have updated grid-column-end to include the width of the world selector component you have added, but we don't need to update grid-row-end for the program block editor, because there was no changes to the rows of overall layout. With this new grid-row-end value, ProgramBlockEditor and PlayAndShare container are fighting for the row 4 -- let's change value of 'grid-row-end' back to what it was before.

`
 grid-row-end: 4;
`

the-t-in-rtf replied:
Right.  Fixed.

chosww commented on src/App.scss:
```diff
@@ -73,84 +104,109 @@ body {
+    height: auto;
+}
+
+.App__playAndShare-background {
```
Now that there's a new column, we need to update grid-column-end of this.

`grid-column-end: 6`

the-t-in-rtf replied:
Fixed.

chosww commented on src/App.scss:
```diff
@@ -73,84 +104,109 @@ body {
+    height: 7rem;
+}
+
+.App__playAndShare-container {
```
For the same reason as its background, this container's grid-column-end needs to be updated as well.

`grid-column-end: 5`

the-t-in-rtf replied:
Fixed.

chosww commented on src/App.scss:
```diff
@@ -73,84 +104,109 @@ body {
 
-.App__program-section {
-    margin-top: 1.5rem;
+.App__world-selector-container {
```
This may change after the latest 286 gets merged in, but world-selector container need same amount of margin-right as the container for ProgramBlockEditor. Let's add the margin.

`margin-right: 0.5rem`

the-t-in-rtf replied:
Done.

chosww commented on src/WorldSelector.js:
```diff
@@ -0,0 +1,85 @@
+}
+
+class WorldSelector extends React.Component<WorldSelectorProps, WorldSelectorState> {
+    constructor (props) {
```
In this implementation, state for world is used to let App know that there was changes in world, so that world state in App gets updated. However, if we just update world state in App directly from onClick/keyDown events from WorldSelector, logics can be simplified and we don't need to store this state in two separate components. Also, world state is needed for other components as well, so keeping this state in one place, at the parent class (App) would be a nice refactoring.

chosww replied:
I suggest removing WorldSelectorState type, constructor, as well as componentDidUpdate method and use WorldSelector as a UI component.

the-t-in-rtf replied:
Makes sense, and done.

chosww commented on src/WorldSelector.js:
```diff
@@ -0,0 +1,85 @@
+                <RobotIcon
+                    aria-label={this.props.intl.formatMessage({id:'WorldSelector.world.default'})}
+                    className={"WorldIcon" + (this.state.world === 'default' ? " WorldIcon--selected" : "") }
+                    onClick={(e: Event) => { this.handleCharacterClick(e, 'default')}}
```
Using arrow function inside of render method is not recommended, because this will create a new function each time the component renders. We can simplify this by assigning value property to each Character buttons, and get this value from handleCharacterClick. 

`onClick={this.handleCharacterClick}`

the-t-in-rtf replied:
Fixed.

chosww commented on src/WorldSelector.js:
```diff
@@ -0,0 +1,85 @@
+                    aria-label={this.props.intl.formatMessage({id:'WorldSelector.world'})}
+                    className="HeaderIcon"
+                />
+                <RobotIcon
```
Let's wrap this icon with a button, and move all the properties in this SVG to it. Also, it would be nice to make this as a toggle button by using aria-pressed property, so users know if the character is selected or not. We also need value property, that's assigned to each character name, in this case, 'default'. Note that we need to use world property from App, with changes to removing state from this component. 

`<button className= {"WorldIcon" + (this.props.world === 'default' ? " WorldIcon--selected" : "") } aria-label={this.props.intl.formatMessage({id:'WorldSelector.world.default}} value='default' onClick={this.handleCharacterClick} onKeyDown={this.handleCharacterKeyDown}> 

`

the-t-in-rtf replied:
Good idea, and done.

chosww commented on src/WorldSelector.js:
```diff
@@ -0,0 +1,85 @@
+        }
+    }
+
+    handleCharacterClick = (event: Event, worldName: WorldName) => {
```
With the suggestions I gave, the handler only needs event parameter, and let's get worldName from button's value property. Also, update directly to App's world state by calling onSelect property. 

`
 this.props.onSelect(event.currentTarget.value)
`

the-t-in-rtf replied:
Done.

chosww commented on src/WorldSelector.js:
```diff
@@ -0,0 +1,85 @@
+
+        if (toggleKeys.indexOf(event.key) !== -1) {
+            event.preventDefault();
+            this.setState({ world: worldName });
```
Same here, update world state directly with onSelect property. 

`
this.props.onSelect(event.currentTarget.value)
`

the-t-in-rtf replied:
Done.

chosww commented on src/WorldSelector.js:
```diff
@@ -0,0 +1,85 @@
+                    className={"WorldIcon" + (this.state.world === 'default' ? " WorldIcon--selected" : "") }
+                    onClick={(e: Event) => { this.handleCharacterClick(e, 'default')}}
+                    onKeyDown={(e: KeyboardEvent) => { this.handleCharacterKeyDown(e, 'default')}}
+                    tabIndex={0}
```
Don't need tabIndex with button component, let's remove this.

the-t-in-rtf replied:
Done.

chosww commented on src/WorldSelector.scss:
```diff
@@ -0,0 +1,40 @@
+    background: #88C3BC;
+}
+
+.WorldSelector button:focus {
```
Focus and hover styles are applied to button component, but these weren't applying because current WorldSelector is not a button.

the-t-in-rtf replied:
I see them now.

chosww commented on src/App.js:
```diff
@@ -578,103 +590,99 @@ export class App extends React.Component<AppProps, AppState> {
+                        </div>
+                    </div>
+                    <div className="App__world-selector-container">
+                        <WorldSelector onSelect={this.handleChangeWorld} />
```
Let's pass world state to WorldSelector.

`
world={this.state.setting.world}
`

the-t-in-rtf replied:
Done.

the-t-in-rtf commented:
Thanks for the feedback, @chosww.  I updated with C2LC-286 and applied the changes you suggested.

I also noticed that the program block editor character was not using the world, and fixed that.

the-t-in-rtf commented:
This should be ready for further review.

chosww commented on src/WorldSelector.js:
```diff
@@ -0,0 +1,81 @@
+                    aria-label={this.props.intl.formatMessage({id:'WorldSelector.world'})}
+                    className="HeaderIcon"
+                />
+                <button
+                    aria-label={this.props.intl.formatMessage({id:'WorldSelector.world.default'})}
+                    aria-pressed={this.props.world === 'default'}
+                    className={"WorldIcon" + (this.props.world === 'default' ? " WorldIcon--selected" : "") }
+                    onClick={this.handleCharacterClick}
+                    onKeyDown={this.handleCharacterKeyDown}
+                    value="default"
+                >
```
As I talked about in Matrix chat, let's use AriaDisablingButton component for these buttons. We need to get condition to disable these buttons from App, let's name the prop disabled.

```javascript
<AriaDisablingButton
    aria-label={this.props.intl.formatMessage({id:'WorldSelector.world.default'})}
    aria-pressed={this.props.world === 'default'}
    className={"WorldIcon" + (this.props.world === 'default' ? " WorldIcon--selected" : "") }
    disabled={this.props.disabled}
    onClick={this.handleCharacterClick}
    onKeyDown={this.handleCharacterKeyDown}
    value="default"
>
```

the-t-in-rtf replied:
Sounds good, I converted all three and updated App.js to disable the world selector when the program is running.

chosww commented on src/App.scss:
```diff
@@ -73,84 +89,135 @@ body {
+    grid-row-start: 1;
+    grid-row-end: 3;
+    margin-right: 0.5rem;
+    width: 6rem;
```
Let's apply the same amount of margin as the scene to the top. 

`
margin-top: 0.5rem;
`

the-t-in-rtf replied:
Done.

chosww commented:
Thanks Tony for the updating the PR, changes you made looks good, just changing the buttons to AriaDisablingButtons would make this PR ready for Simon to review.

the-t-in-rtf commented:
Thanks, @chosww.  I made the changes you suggested.

sbates-idrc commented:
@the-t-in-rtf please resolve conflicts with `develop-0.7` and then I will merge
